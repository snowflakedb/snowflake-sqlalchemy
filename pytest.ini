# pytest.ini
[pytest]
# There are majorly two issues with the sqlalchemy test suites if we want to run it with other snowflake specific tests:
# 1. If including the test_suite.py, only tests in test_suite.py will be collected, other dialect tests are not collected
# due to "from sqlalchemy.testing.plugin.pytestplugin import".
# 2. Putting test_suites.py and related conftest config under tests root dir would make pytest fail to collect cfg
# information such as requirements, and customized session db config does not work as expected (schema not set).
# Thus, we ignore the path for now, and give instructions to run the test suites in standalone directory.
addopts = --ignore=tests/sqlalchemy_test_suite
