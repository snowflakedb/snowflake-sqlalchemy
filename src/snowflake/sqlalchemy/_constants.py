#
# Copyright (c) 2012-2022 Snowflake Computing Inc. All rights reserved.
#
import pkg_resources

# parameters needed for usage tracking
PARAM_APPLICATION = "application"
PARAM_INTERNAL_APPLICATION_NAME = "internal_application_name"
PARAM_INTERNAL_APPLICATION_VERSION = "internal_application_version"

APPLICATION_NAME = "SnowflakeSQLAlchemy"
SNOWFLAKE_SQLALCHEMY_VERSION = pkg_resources.get_distribution(
    "snowflake-sqlalchemy"
).version
