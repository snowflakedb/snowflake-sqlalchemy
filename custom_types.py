#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# Copyright (c) 2012-2019 Snowflake Computing Inc. All right reserved.
#

import sqlalchemy.types as sqltypes

TEXT = sqltypes.VARCHAR
CHARACTER = sqltypes.CHAR
DEC = sqltypes.DECIMAL
DOUBLE = sqltypes.FLOAT
FIXED = sqltypes.DECIMAL
NUMBER = sqltypes.DECIMAL
BYTEINT = sqltypes.SMALLINT
STRING = sqltypes.VARCHAR
TINYINT  = sqltypes.SMALLINT
VARBINARY = sqltypes.BINARY

class VARIANT(sqltypes.TypeEngine):
    __visit_name__ = 'VARIANT'


class OBJECT(sqltypes.TypeEngine):
    __visit_name__ = 'OBJECT'


class ARRAY(sqltypes.TypeEngine):
    __visit_name__ = 'ARRAY'


class TIMESTAMP_TZ(sqltypes.TIMESTAMP):
    __visit_name__ = 'TIMESTAMP_TZ'


class TIMESTAMP_LTZ(sqltypes.TIMESTAMP):
    __visit_name__ = 'TIMESTAMP_LTZ'


class TIMESTAMP_NTZ(sqltypes.TIMESTAMP):
    __visit_name__ = 'TIMESTAMP_NTZ'
