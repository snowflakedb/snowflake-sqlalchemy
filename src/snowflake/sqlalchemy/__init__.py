#
# Copyright (c) 2012-2023 Snowflake Computing Inc. All rights reserved.
#

import sys

if sys.version_info < (3, 8):
    import importlib_metadata
else:
    import importlib.metadata as importlib_metadata

from sqlalchemy.types import (  # noqa
    BIGINT,
    BINARY,
    BOOLEAN,
    CHAR,
    DATE,
    DATETIME,
    DECIMAL,
    FLOAT,
    INT,
    INTEGER,
    REAL,
    SMALLINT,
    TIME,
    TIMESTAMP,
    VARCHAR,
)

from . import base, snowdialect  # noqa
from .custom_commands import (  # noqa
    AWSBucket,
    AzureContainer,
    CopyFormatter,
    CopyIntoStorage,
    CreateFileFormat,
    CreateStage,
    CSVFormatter,
    ExternalStage,
    JSONFormatter,
    MergeInto,
    PARQUETFormatter,
)
from .custom_types import (  # noqa
    ARRAY,
    BYTEINT,
    CHARACTER,
    DEC,
    DOUBLE,
    FIXED,
    GEOGRAPHY,
    GEOMETRY,
    MAP,
    NUMBER,
    OBJECT,
    STRING,
    TEXT,
    TIMESTAMP_LTZ,
    TIMESTAMP_NTZ,
    TIMESTAMP_TZ,
    TINYINT,
    VARBINARY,
    VARIANT,
)
from .sql.custom_schema import (  # noqa
    DynamicTable,
    HybridTable,
    IcebergTable,
    SnowflakeTable,
)
from .sql.custom_schema.options import (  # noqa
    AsQueryOption,
    ClusterByOption,
    IdentifierOption,
    KeywordOption,
    LiteralOption,
    SnowflakeKeyword,
    TableOptionKey,
    TargetLagOption,
    TimeUnit,
)
from .util import _url as URL  # noqa

base.dialect = dialect = snowdialect.dialect

__version__ = importlib_metadata.version("snowflake-sqlalchemy")

_custom_types = (
    "BIGINT",
    "BINARY",
    "BOOLEAN",
    "CHAR",
    "DATE",
    "DATETIME",
    "DECIMAL",
    "FLOAT",
    "INT",
    "INTEGER",
    "REAL",
    "SMALLINT",
    "TIME",
    "TIMESTAMP",
    "URL",
    "VARCHAR",
    "ARRAY",
    "BYTEINT",
    "CHARACTER",
    "DEC",
    "DOUBLE",
    "FIXED",
    "GEOGRAPHY",
    "GEOMETRY",
    "OBJECT",
    "NUMBER",
    "STRING",
    "TEXT",
    "TIMESTAMP_LTZ",
    "TIMESTAMP_TZ",
    "TIMESTAMP_NTZ",
    "TINYINT",
    "VARBINARY",
    "VARIANT",
    "MAP",
)

_custom_commands = (
    "MergeInto",
    "CSVFormatter",
    "JSONFormatter",
    "PARQUETFormatter",
    "CopyFormatter",
    "CopyIntoStorage",
    "AWSBucket",
    "AzureContainer",
    "ExternalStage",
    "CreateStage",
    "CreateFileFormat",
)

_custom_tables = ("HybridTable", "DynamicTable", "IcebergTable", "SnowflakeTable")

_custom_table_options = (
    "AsQueryOption",
    "TargetLagOption",
    "LiteralOption",
    "IdentifierOption",
    "KeywordOption",
    "ClusterByOption",
)

_enums = (
    "TimeUnit",
    "TableOptionKey",
    "SnowflakeKeyword",
)
__all__ = (
    *_custom_types,
    *_custom_commands,
    *_custom_tables,
    *_custom_table_options,
    *_enums,
)
