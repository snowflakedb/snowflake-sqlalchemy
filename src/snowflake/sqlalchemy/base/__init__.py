# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from sqlalchemy.schema import Table
from sqlalchemy.sql import functions

from ..functions import flatten
from .compiler.snowflake_compiler import SnowflakeCompiler
from .compiler.snowflake_ddl_compiler import SnowflakeDDLCompiler
from .compiler.snowflake_identifier_preparer import SnowflakeIdentifierPreparer
from .compiler.snowflake_type_compiler import SnowflakeTypeCompiler
from .constants import RESERVED_WORDS
from .snowflake_execution_context import SnowflakeExecutionContext
from .snowflake_orm_select_compile_state import SnowflakeORMSelectCompileState
from .snowflake_select_state import SnowflakeSelectState

# The __all__ list is used to maintain backward compatibility with previous versions of the package.
# After splitting a large file into a module, __all__ explicitly defines the public API, ensuring
# that existing imports from the original file still work as expected. This approach prevents
# breaking changes by controlling which functions, classes, and variables are exposed when
# the module is imported.
__all__ = [
    "SnowflakeDDLCompiler",
    "SnowflakeCompiler",
    "SnowflakeIdentifierPreparer",
    "SnowflakeTypeCompiler",
    "SnowflakeExecutionContext",
    "SnowflakeORMSelectCompileState",
    "SnowflakeSelectState",
    "RESERVED_WORDS",
]


construct_arguments = [(Table, {"clusterby": None})]

functions.register_function("flatten", flatten)
