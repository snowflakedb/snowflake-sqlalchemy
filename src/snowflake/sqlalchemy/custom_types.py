#
# Copyright (c) 2012-2022 Snowflake Computing Inc. All rights reserved.
#
import datetime
import decimal
import re

import sqlalchemy.types as sqltypes
import sqlalchemy.util as util

TEXT = sqltypes.VARCHAR
CHARACTER = sqltypes.CHAR
DEC = sqltypes.DECIMAL
DOUBLE = sqltypes.FLOAT
FIXED = sqltypes.DECIMAL
NUMBER = sqltypes.DECIMAL
BYTEINT = sqltypes.SMALLINT
STRING = sqltypes.VARCHAR
TINYINT = sqltypes.SMALLINT
VARBINARY = sqltypes.BINARY


def _process_float(value):
    if value == float("inf"):
        return "inf"
    elif value == float("-inf"):
        return "-inf"
    elif value is not None:
        return float(value)
    return value


class SnowflakeType(sqltypes.TypeEngine):
    def _default_dialect(self):
        # Get around circular import
        return __import__("snowflake.sqlalchemy").sqlalchemy.dialect()


class VARIANT(SnowflakeType):
    __visit_name__ = "VARIANT"


class OBJECT(SnowflakeType):
    __visit_name__ = "OBJECT"


class ARRAY(SnowflakeType):
    __visit_name__ = "ARRAY"


class TIMESTAMP_TZ(SnowflakeType):
    __visit_name__ = "TIMESTAMP_TZ"


class TIMESTAMP_LTZ(SnowflakeType):
    __visit_name__ = "TIMESTAMP_LTZ"


class TIMESTAMP_NTZ(SnowflakeType):
    __visit_name__ = "TIMESTAMP_NTZ"


class GEOGRAPHY(SnowflakeType):
    __visit_name__ = "GEOGRAPHY"


class _CUSTOM_Date(SnowflakeType, sqltypes.Date):
    def literal_processor(self, dialect):
        def process(value):
            if value is not None:
                return f"'{value.isoformat()}'"

        return process

    _reg = re.compile(r"(\d+)-(\d+)-(\d+)")

    def result_processor(self, dialect, coltype):
        def process(value):
            if isinstance(value, str):
                m = self._reg.match(value)
                if not m:
                    raise ValueError(f"could not parse {value!r} as a date value")
                return datetime.date(*[int(x or 0) for x in m.groups()])
            else:
                return value

        return process


class _CUSTOM_DateTime(SnowflakeType, sqltypes.DateTime):
    def literal_processor(self, dialect):
        def process(value):
            if value is not None:
                datetime_str = value.isoformat(" ", timespec="microseconds")
                return f"'{datetime_str}'"

        return process

    _reg = re.compile(r"(\d+)-(\d+)-(\d+) (\d+):(\d+):(\d+)(?:\.(\d{0,6}))?")

    def result_processor(self, dialect, coltype):
        def process(value):
            if isinstance(value, str):
                m = self._reg.match(value)
                if not m:
                    raise ValueError(f"could not parse {value!r} as a datetime value")
                return datetime.datetime(*[int(x or 0) for x in m.groups()])
            else:
                return value

        return process


class _CUSTOM_Time(SnowflakeType, sqltypes.Time):
    def literal_processor(self, dialect):
        def process(value):
            if value is not None:
                time_str = value.isoformat(timespec="microseconds")
                return f"'{time_str}'"

        return process

    _reg = re.compile(r"(\d+):(\d+):(\d+)(?:\.(\d{0,6}))?")

    def result_processor(self, dialect, coltype):
        def process(value):
            if isinstance(value, str):
                m = self._reg.match(value)
                if not m:
                    raise ValueError(f"could not parse {value!r} as a time value")
                return datetime.time(*[int(x or 0) for x in m.groups()])
            else:
                return value

        return process


class _CUSTOM_Float(SnowflakeType, sqltypes.Float):
    def bind_processor(self, dialect):
        return _process_float


class _CUSTOM_DECIMAL(SnowflakeType, sqltypes.DECIMAL):
    @util.memoized_property
    def _type_affinity(self):
        return sqltypes.INTEGER if self.scale == 0 else sqltypes.DECIMAL


class _CUSTOM_Numeric(SnowflakeType, sqltypes.Numeric):
    def result_processor(self, dialect, coltype):
        if self.asdecimal:

            def process(value):
                if value:
                    return decimal.Decimal(value)
                else:
                    return None

            return process
        else:
            return _process_float
