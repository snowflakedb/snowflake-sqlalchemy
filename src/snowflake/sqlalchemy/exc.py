#
# Copyright (c) 2012-2023 Snowflake Computing Inc. All rights reserved.

from typing import List

from sqlalchemy.exc import ArgumentError


class NoPrimaryKeyError(ArgumentError):
    def __init__(self, target: str):
        super().__init__(f"Table {target} required primary key.")


class UnsupportedPrimaryKeysAndForeignKeysError(ArgumentError):
    def __init__(self, target: str):
        super().__init__(f"Primary key and foreign keys are not supported in {target}.")


class RequiredParametersNotProvidedError(ArgumentError):
    def __init__(self, target: str, parameters: List[str]):
        super().__init__(
            f"{target} requires the following parameters: %s." % ", ".join(parameters)
        )


class UnexpectedTableOptionKeyError(ArgumentError):
    def __init__(self, expected: str, actual: str):
        super().__init__(f"Expected table option {expected} but got {actual}.")


class OptionKeyNotProvidedError(ArgumentError):
    def __init__(self, target: str):
        super().__init__(
            f"Expected option key in {target} option but got NoneType instead."
        )


class UnexpectedOptionParameterTypeError(ArgumentError):
    def __init__(self, parameter_name: str, target: str, types: List[str]):
        super().__init__(
            f"Parameter {parameter_name} of {target} requires to be one"
            f" of following types: {', '.join(types)}."
        )


class CustomOptionsAreOnlySupportedOnSnowflakeTables(ArgumentError):
    def __init__(self):
        super().__init__(
            "Identifier, Literal, TargetLag and other custom options are only supported on Snowflake tables."
        )


class UnexpectedOptionTypeError(ArgumentError):
    def __init__(self, options: List[str]):
        super().__init__(
            f"The following options are either unsupported or should be defined using a Snowflake table: {', '.join(options)}."
        )


class InvalidTableParameterTypeError(ArgumentError):
    def __init__(self, name: str, input_type: str, expected_types: List[str]):
        expected_types_str = "', '".join(expected_types)
        super().__init__(
            f"Invalid parameter type '{input_type}' provided for '{name}'. "
            f"Expected one of the following types: '{expected_types_str}'.\n"
        )


class MultipleErrors(ArgumentError):
    def __init__(self, errors):
        self.errors = errors

    def __str__(self):
        return "".join(str(e) for e in self.errors)


class StructuredTypeNotSupportedInTableColumnsError(ArgumentError):
    def __init__(self, table_type: str, table_name: str, column_name: str):
        super().__init__(
            f"Column '{column_name}' is of a structured type, which is only supported on Iceberg tables. "
            f"The table '{table_name}' is of type '{table_type}', not Iceberg."
        )
