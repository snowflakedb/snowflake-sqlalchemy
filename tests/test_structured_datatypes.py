#
# Copyright (c) 2012-2023 Snowflake Computing Inc. All rights reserved.
#

import pytest
from sqlalchemy import (
    Column,
    Integer,
    MetaData,
    Sequence,
    Table,
    cast,
    exc,
    inspect,
    text,
)
from sqlalchemy.orm import Session, declarative_base
from sqlalchemy.sql import select
from sqlalchemy.sql.ddl import CreateTable

from snowflake.sqlalchemy import NUMBER, IcebergTable, SnowflakeTable
from snowflake.sqlalchemy.custom_types import MAP, TEXT
from snowflake.sqlalchemy.exc import StructuredTypeNotSupportedInTableColumnsError


def test_compile_table_with_cluster_by_with_expression(sql_compiler, snapshot):
    metadata = MetaData()
    user_table = Table(
        "clustered_user",
        metadata,
        Column("Id", Integer, primary_key=True),
        Column("name", MAP(NUMBER(), TEXT())),
    )

    create_table = CreateTable(user_table)

    assert sql_compiler(create_table) == snapshot


@pytest.mark.requires_external_volume
def test_create_table_structured_datatypes(
    engine_testaccount, external_volume, base_location
):
    metadata = MetaData()
    table_name = "test_map0"
    test_map = IcebergTable(
        table_name,
        metadata,
        Column("id", Integer, primary_key=True),
        Column("map_id", MAP(NUMBER(10, 0), TEXT())),
        external_volume=external_volume,
        base_location=base_location,
    )
    metadata.create_all(engine_testaccount)
    try:
        assert test_map is not None
    finally:
        test_map.drop(engine_testaccount)


@pytest.mark.requires_external_volume
def test_insert_map(engine_testaccount, external_volume, base_location, snapshot):
    metadata = MetaData()
    table_name = "test_insert_map"
    test_map = IcebergTable(
        table_name,
        metadata,
        Column("id", Integer, primary_key=True),
        Column("map_id", MAP(NUMBER(10, 0), TEXT())),
        external_volume=external_volume,
        base_location=base_location,
    )
    metadata.create_all(engine_testaccount)

    try:
        with engine_testaccount.connect() as conn:
            slt = select(
                1,
                cast(
                    text("{'100':'item1', '200':'item2'}"), MAP(NUMBER(10, 0), TEXT())
                ),
            )
            ins = test_map.insert().from_select(["id", "map_id"], slt)
            conn.execute(ins)

            results = conn.execute(test_map.select())
            data = results.fetchmany()
            results.close()
            snapshot.assert_match(data)
    finally:
        test_map.drop(engine_testaccount)


@pytest.mark.requires_external_volume
@pytest.mark.parametrize(
    "structured_type",
    [
        MAP(NUMBER(10, 0), TEXT()),
        MAP(NUMBER(10, 0), MAP(NUMBER(10, 0), TEXT())),
    ],
)
def test_inspect_structured_data_types(
    engine_testaccount, external_volume, base_location, snapshot, structured_type
):
    metadata = MetaData()
    table_name = "test_st_types"
    test_map = IcebergTable(
        table_name,
        metadata,
        Column("id", Integer, primary_key=True),
        Column("map_id", structured_type),
        external_volume=external_volume,
        base_location=base_location,
    )
    metadata.create_all(engine_testaccount)

    try:
        inspecter = inspect(engine_testaccount)
        columns = inspecter.get_columns(table_name)

        assert isinstance(columns[0]["type"], NUMBER)
        assert isinstance(columns[1]["type"], MAP)
        assert columns == snapshot

    finally:
        test_map.drop(engine_testaccount)


@pytest.mark.requires_external_volume
@pytest.mark.parametrize(
    "structured_type",
    [
        "MAP(NUMBER(10, 0), VARCHAR)",
        "MAP(NUMBER(10, 0), MAP(NUMBER(10, 0), VARCHAR))",
    ],
)
def test_reflect_structured_data_types(
    engine_testaccount,
    external_volume,
    base_location,
    snapshot,
    structured_type,
    sql_compiler,
):
    metadata = MetaData()
    table_name = "test_reflect_st_types"
    create_table_sql = f"""
CREATE OR REPLACE ICEBERG TABLE {table_name} (
       id number(38,0) primary key,
       map_id {structured_type})
CATALOG = 'SNOWFLAKE'
EXTERNAL_VOLUME = '{external_volume}'
BASE_LOCATION = '{base_location}';
      """

    with engine_testaccount.connect() as connection:
        connection.exec_driver_sql(create_table_sql)

    iceberg_table = IcebergTable(table_name, metadata, autoload_with=engine_testaccount)
    constraint = iceberg_table.constraints.pop()
    constraint.name = "constraint_name"
    iceberg_table.constraints.add(constraint)

    try:
        with engine_testaccount.connect():
            value = CreateTable(iceberg_table)

            actual = sql_compiler(value)

            assert actual == snapshot

    finally:
        metadata.drop_all(engine_testaccount)


@pytest.mark.requires_external_volume
def test_insert_map_orm(
    sql_compiler, external_volume, base_location, engine_testaccount, snapshot
):
    Base = declarative_base()
    session = Session(bind=engine_testaccount)

    class TestIcebergTableOrm(Base):
        __tablename__ = "test_iceberg_table_orm"

        @classmethod
        def __table_cls__(cls, name, metadata, *arg, **kw):
            return IcebergTable(name, metadata, *arg, **kw)

        __table_args__ = {
            "external_volume": external_volume,
            "base_location": base_location,
        }

        id = Column(Integer, Sequence("user_id_seq"), primary_key=True)
        map_id = Column(MAP(NUMBER(10, 0), TEXT()))

        def __repr__(self):
            return f"({self.id!r}, {self.name!r})"

    Base.metadata.create_all(engine_testaccount)

    try:
        cast_expr = cast(
            text("{'100':'item1', '200':'item2'}"), MAP(NUMBER(10, 0), TEXT())
        )
        instance = TestIcebergTableOrm(id=0, map_id=cast_expr)
        session.add(instance)
        with pytest.raises(exc.ProgrammingError) as programming_error:
            session.commit()
        # TODO: Support variant in insert statement
        assert str(programming_error.value.orig) == snapshot
    finally:
        Base.metadata.drop_all(engine_testaccount)


def test_snowflake_tables_with_structured_types(sql_compiler):
    metadata = MetaData()
    with pytest.raises(
        StructuredTypeNotSupportedInTableColumnsError
    ) as programming_error:
        SnowflakeTable(
            "clustered_user",
            metadata,
            Column("Id", Integer, primary_key=True),
            Column("name", MAP(NUMBER(10, 0), TEXT())),
        )
    assert programming_error is not None


@pytest.mark.requires_external_volume
def test_select_map_orm(engine_testaccount, external_volume, base_location, snapshot):
    metadata = MetaData()
    table_name = "test_select_map_orm"
    test_map = IcebergTable(
        table_name,
        metadata,
        Column("id", Integer, primary_key=True),
        Column("map_id", MAP(NUMBER(10, 0), TEXT())),
        external_volume=external_volume,
        base_location=base_location,
    )
    metadata.create_all(engine_testaccount)

    with engine_testaccount.connect() as conn:
        slt1 = select(
            2,
            cast(text("{'100':'item1', '200':'item2'}"), MAP(NUMBER(10, 0), TEXT())),
        )
        slt2 = select(
            1,
            cast(text("{'100':'item1', '200':'item2'}"), MAP(NUMBER(10, 0), TEXT())),
        ).union_all(slt1)
        ins = test_map.insert().from_select(["id", "map_id"], slt2)
        conn.execute(ins)
        conn.commit()

    Base = declarative_base()
    session = Session(bind=engine_testaccount)

    class TestIcebergTableOrm(Base):
        __table__ = test_map

        def __repr__(self):
            return f"({self.id!r}, {self.map_id!r})"

    try:
        data = session.query(TestIcebergTableOrm).all()
        snapshot.assert_match(data)
    finally:
        test_map.drop(engine_testaccount)
