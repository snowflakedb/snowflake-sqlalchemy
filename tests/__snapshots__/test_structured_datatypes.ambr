# serializer version: 1
# name: test_compile_table_with_cluster_by_with_expression
  'CREATE TABLE clustered_user (\t"Id" INTEGER NOT NULL AUTOINCREMENT, \tname MAP(DECIMAL, VARCHAR), \tPRIMARY KEY ("Id"))'
# ---
# name: test_compile_table_with_double_map
  'CREATE TABLE clustered_user (\t"Id" INTEGER NOT NULL AUTOINCREMENT, \tname MAP(DECIMAL, MAP(DECIMAL, VARCHAR)), \tPRIMARY KEY ("Id"))'
# ---
# name: test_insert_map
  list([
    (1, '{\n  "100": "item1",\n  "200": "item2"\n}'),
  ])
# ---
# name: test_insert_map_orm
  '''
  002014 (22000): SQL compilation error:
  Invalid expression [CAST(OBJECT_CONSTRUCT('100', 'item1', '200', 'item2') AS MAP(NUMBER(10,0), VARCHAR(16777216)))] in VALUES clause
  '''
# ---
# name: test_inspect_structured_data_types[structured_type0]
  list([
    dict({
      'autoincrement': True,
      'comment': None,
      'default': None,
      'identity': dict({
        'increment': 1,
        'start': 1,
      }),
      'name': 'id',
      'nullable': False,
      'primary_key': True,
      'type': _CUSTOM_DECIMAL(precision=10, scale=0),
    }),
    dict({
      'autoincrement': False,
      'comment': None,
      'default': None,
      'name': 'map_id',
      'nullable': True,
      'primary_key': False,
      'type': MAP(_CUSTOM_DECIMAL(precision=10, scale=0), VARCHAR(length=16777216)),
    }),
  ])
# ---
# name: test_inspect_structured_data_types[structured_type1]
  list([
    dict({
      'autoincrement': True,
      'comment': None,
      'default': None,
      'identity': dict({
        'increment': 1,
        'start': 1,
      }),
      'name': 'id',
      'nullable': False,
      'primary_key': True,
      'type': _CUSTOM_DECIMAL(precision=10, scale=0),
    }),
    dict({
      'autoincrement': False,
      'comment': None,
      'default': None,
      'name': 'map_id',
      'nullable': True,
      'primary_key': False,
      'type': MAP(_CUSTOM_DECIMAL(precision=10, scale=0), MAP(_CUSTOM_DECIMAL(precision=10, scale=0), VARCHAR(length=16777216))),
    }),
  ])
# ---
# name: test_reflect_structured_data_types[MAP(NUMBER(10, 0), MAP(NUMBER(10, 0), VARCHAR))]
  "CREATE ICEBERG TABLE test_reflect_st_types (\tid DECIMAL(38, 0) NOT NULL, \tmap_id MAP(DECIMAL(10, 0), MAP(DECIMAL(10, 0), VARCHAR(16777216))), \tCONSTRAINT constraint_name PRIMARY KEY (id))\tCATALOG = 'SNOWFLAKE'"
# ---
# name: test_reflect_structured_data_types[MAP(NUMBER(10, 0), VARCHAR)]
  "CREATE ICEBERG TABLE test_reflect_st_types (\tid DECIMAL(38, 0) NOT NULL, \tmap_id MAP(DECIMAL(10, 0), VARCHAR(16777216)), \tCONSTRAINT constraint_name PRIMARY KEY (id))\tCATALOG = 'SNOWFLAKE'"
# ---
# name: test_select_map_orm
  list([
    (1, '{\n  "100": "item1",\n  "200": "item2"\n}'),
    (2, '{\n  "100": "item1",\n  "200": "item2"\n}'),
  ])
# ---
# name: test_select_map_orm.1
  list([
  ])
# ---
# name: test_select_map_orm.2
  list([
  ])
# ---
